### PR TITLE
refactor: wss revert has been reimplemented to utilize socket.io acknowledgement.

### PR DESCRIPTION
**Pull request by mmoehabb**

### CODE REVIEW

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-8zl46b3a9-litespace-core.vercel.app) 
🚀 [Luna](https://luna-r1v8ceii2-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-inzfyzi4r-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-94r9ts2wa-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-goqm02xy1-litespace-core.vercel.app) 
🚀 [Luna](https://luna-19hxi3dyh-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-h1a95acik-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-50cf89sj3-litespace-core.vercel.app)

File: services/server/src/wss/handlers/base.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { IUser, Wss } from "@litespace/types";
+import { AcknowledgeCallback, AcknowledgePayload } from "@litespace/types/dist/esm/wss";
```
Comment (neuodev) [inline services/server/src/wss/handlers/base.ts:2]: This is not a valid import.

File: packages/types/src/wss.ts [hunk 2]
```diff
@@ -72,47 +71,31 @@ export enum Room {
...
-      reason: string;
-    }>;
-
-type EventCallback<T> = (arg: T) => Promise<void> | void;
+export type AcknowledgePayload = {
+  code: AcknowledgeCodes;
+  message?: string;
+}
+
+export type AcknowledgeCallback = (payload?: AcknowledgePayload) => any;
```
Comment (neuodev) [inline packages/types/src/wss.ts:90]:
```suggestion
export type AcknowledgeCallback = (payload?: AcknowledgePayload) => void;
```

File: services/server/src/wss/handlers/base.ts [hunk 3]
```diff
@@ -19,7 +20,7 @@ export abstract class WssHandler {
     this.socket.broadcast.to(room).emit(event, ...data);
   }
 
-  revert(payload: Wss.RevertPayload) {
-    this.socket.emit(Wss.ServerEvent.Revert, payload);
+  protected must(callback: AcknowledgeCallback | undefined, payload: AcknowledgePayload): void {
+    return callback ? callback(payload) : console.error("Socket.io callback is not defined!");
   }
```
Comment (neuodev) [inline services/server/src/wss/handlers/base.ts:25]: No need to throw/log an error, The frontend might not be interested in cacthing errors.

```suggestion
  protected call(callback: AcknowledgeCallback | undefined, payload: AcknowledgePayload): void {
    if (!callback) return;
    return callback(payload);
  }
```

File: services/server/src/wss/handlers/connection.ts [hunk 4]
```diff
@@ -58,15 +58,20 @@ export class Connection extends WssHandler {
...
   }
 
   private async emitServerStats() {
     background.on("message", (message: PartentPortMessage) => {
       if (message.type === PartentPortMessageType.Stats)
-        return this.socket.emit(Wss.ServerEvent.ServerStats, message.stats);
+        return this.socket.emit(
+          Wss.ServerEvent.ServerStats, 
+          message.stats,
+          () => {}
```
Comment (neuodev) [inline services/server/src/wss/handlers/connection.ts:73]: No need to have this empty callback. 


```suggestion
```

File: services/server/src/wss/handlers/connection.ts [hunk 5]
```diff
@@ -58,15 +58,20 @@ export class Connection extends WssHandler {
       this.broadcast(
         Wss.ServerEvent.UserStatusChanged,
         room.toString(),
-        { online }
+        { online },
+        () => {}
```
Comment (neuodev) [inline services/server/src/wss/handlers/connection.ts:62]:
```suggestion
```

File: services/server/src/wss/handlers/connection.ts [hunk 6]
```diff
@@ -124,8 +129,10 @@ export class Connection extends WssHandler {
...
       .to(asCallRoomId(callId))
-      .emit(Wss.ServerEvent.MemberLeftCall, {
-        userId: user.id,
-      });
+      .emit(
+        Wss.ServerEvent.MemberLeftCall, 
+        { userId: user.id },
+        () => {}
+      );
   }
```
Comment (neuodev) [inline services/server/src/wss/handlers/connection.ts:137]:
```suggestion
      .emit(Wss.ServerEvent.MemberLeftCall, { userId: user.id });
```

File: services/server/src/wss/handlers/inputDevices.ts [hunk 7]
```diff
@@ -31,7 +29,7 @@ export class InputDevices extends WssHandler {
       this.broadcast(Wss.ServerEvent.CameraToggled, call.toString(), {
         user: user.id,
         camera,
-      });
+      }, () => {});
```
Comment (neuodev) [inline services/server/src/wss/handlers/inputDevices.ts:32]: Please remove these empty callbacks.

File: packages/types/src/wss.ts [hunk 8]
```diff
@@ -72,47 +71,31 @@ export enum Room {
   ServerStats = "ServerStats",
 }
 
-export enum RevertErrorCode {
+export enum AcknowledgeCodes {
```
Comment (neuodev) [inline packages/types/src/wss.ts:74]: Enums should be singular. In this case it should be `AcknowledgeCode`

**Review by neuodev — CHANGES_REQUESTED** [review]

File: services/server/src/wss/handlers/base.ts (hunk 3 above)
Comment (mmoehabb) [reply services/server/src/wss/handlers/base.ts:25]: This is for debugging in the backend... no thing to do with the frontned. Calling `this.must` or `this.call` without any reflection will be mysterious and puzzles the user (developer).

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-cpbzpyage-litespace-core.vercel.app) 
🚀 [Luna](https://luna-4zeglhid7-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-ksscmlrq8-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-2znxguwkk-litespace-core.vercel.app)

File: services/server/src/wss/handlers/base.ts (hunk 3 above)
Comment (neuodev) [reply services/server/src/wss/handlers/base.ts]: From the first look, it is not clear that it is used for debugging. If you want to warn the developer not to do this, then you should throw an error instead.

Comment (mmoehabb) [reply services/server/src/wss/handlers/base.ts]: It's completely normal for callback to be undefined; in other word, it's not an error to not pass the callback parameter. That's why I considered using the log.

Comment (neuodev) [reply services/server/src/wss/handlers/base.ts]: Sounds good.

**Review by neuodev — APPROVED** [review]
LGTM! Great work @mmoehabb !

**Pull request merged**